### PR TITLE
refacto/feat: create a layer/source from data loaded by the user

### PR DESCRIPTION
**Pull request by zarov**
## Description

Main goal of the PR: create a Layer/Source directly from data loaded 
by the user. To achieve this, it is mainly `DataSourceProvider` and
`FileSource` that have been rewritten. So now we have three way of 
receiving data:
- by fetching it, parsing it and converting it in the provider
- by giving fetched data to the layer and parsing it and converting it
  in the provider
- by giving parsed data to the layer and converting it in the provider

And thus FileSource can be simplified drastically and extended to
receive ANY type of data, not only geojson, kml and gpx (see #1032)

Documentation and examples have been updated, and I tried to not 
introduce breaking changes, but I may have miss some.

## Motivation and Context

People have asked about it in issue (#892), offline and I need this 
working for another thing. It is also the occasion to clean `FileSource` 
and `DataSourceProvider` that were too messy imho.

## Problems

I still got two problems, that maybe someone can help with:
- the first one is visible in example `globe_vector`, the GPX path and 
the geojson texture seems to overlap, but I don't understand why
- in example `globe_geojson_to3D`, I can't seem to make it work too

NOTE: this PR needs #1032 to be merged first

### CODE REVIEW

**Comment by zarov** [comment]
Something else I would change: as you can see, you can use `source.fetchedData` and `source.parsedData` to add new data that you have already fetched or parsed. To avoid complexity for the user, I was thinking to maybe also add `data` when declaring the source, so you could have:

```js
Fetcher.json('foo.geojson').then((fetchedData => {
    var source = new FileSource('source', {
        data: fetchedData,
    });
...
``` 
or

```js
Fetcher.json('foo.geojson').then(GeoJsonParser.parse)
    .then((parsedData => {
        var source = new FileSource('source', {
            data: parsedData,
        });
...
``` 

WDYT ?

**Comment by zarov** [comment]
A quick update: `globe_geojson_to3D` seems to work partially, it seems there is a problem in `FeatureProcessing`. Needs more investigation unfortunately.

**Comment by zarov** [comment]
Update: the first problem (`globe_vector`) has been explained in #1040, so I'm considering it as not blocking this PR anymore. The second problem (`globe_geojson_to3d`) seems fixed.

File: examples/globe_geojson_to3D.html
Comment (gchoqueux) [inline examples/globe_geojson_to3D.html]: it is not up to the user to specify the target projection. it must be `view.referenceCrs`?

File: examples/globe_geojson_to3D.html
Comment (gchoqueux) [inline examples/globe_geojson_to3D.html]: it will be necessary to make more friendly: add specofic `Source` by format. This is **Breaking change**? it's necessary to add parser?

File: examples/globe_vector.html
Comment (gchoqueux) [inline examples/globe_vector.html]: it's not up to the user to specify the target projection, this should be done by itowns, when adding the layer.

File: examples/globe_vector.html
Comment (gchoqueux) [inline examples/globe_vector.html]: in this case, the projection is necessary?

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js:194]: nice

File: src/Process/FeatureProcessing.js
Comment (gchoqueux) [inline src/Process/FeatureProcessing.js:116]: could we use the global extent of `parsedData`?

File: src/Provider/DataSourceProvider.js
Comment (gchoqueux) [inline src/Provider/DataSourceProvider.js]: the source shouldn't decide in the options parsing, it's rather the layer that influences the options.

File: src/Source/FileSource.js
Comment (gchoqueux) [inline src/Source/FileSource.js:7]: a good suppression

File: src/Source/FileSource.js
Comment (gchoqueux) [inline src/Source/FileSource.js:122]: could you verify if `source.url` is undefined so `fetchedData` or `parsedData` are defined?

**Review by gchoqueux — CHANGES_REQUESTED** [review]
thanks, it's a good enhancement

File: examples/globe_geojson_to3D.html
Comment (zarov) [reply examples/globe_geojson_to3D.html]: Yes I agree, we can add some new Sources. This is not a breaking change, as `format` is still read. In fact, we don't really need to have the parser specified here.

File: examples/globe_geojson_to3D.html
Comment (zarov) [reply examples/globe_geojson_to3D.html]: Ok I'm looking into changing that

File: examples/globe_vector.html
Comment (zarov) [reply examples/globe_vector.html]: No, but the previous version of `FileSource` has `projection` mandatory, so I kept it like this. But we can change it if you want.

File: src/Process/FeatureProcessing.js
Comment (zarov) [reply src/Process/FeatureProcessing.js:116]: It is the case below ? I don't understand sorry

File: src/Provider/DataSourceProvider.js
Comment (zarov) [reply src/Provider/DataSourceProvider.js]: Ok

File: src/Source/FileSource.js
Comment (zarov) [reply src/Source/FileSource.js:122]: Yes good idea

File: examples/globe_geojson_to3D.html
Comment (zarov) [reply examples/globe_geojson_to3D.html]: Removed, because unnecessary

File: examples/globe_geojson_to3D.html
Comment (zarov) [reply examples/globe_geojson_to3D.html]: Removed to avoid the confusion

File: examples/globe_vector.html
Comment (zarov) [reply examples/globe_vector.html]: Removed

File: examples/globe_vector.html
Comment (zarov) [reply examples/globe_vector.html]: Removed because the `FileSource` now reads `parsedData` if present.

File: src/Process/FeatureProcessing.js
Comment (zarov) [reply src/Process/FeatureProcessing.js:116]: Factorized

File: src/Provider/DataSourceProvider.js
Comment (zarov) [reply src/Provider/DataSourceProvider.js]: Removed

File: src/Source/FileSource.js
Comment (zarov) [reply src/Source/FileSource.js:122]: Done

**Comment by gchoqueux** [comment]
in `/examples/globe_vector.html`, there are 2 problems:
* no line with gpx;
* **ariege** json can be displayed halfway;

**Comment by zarov** [comment]
* I'm working on the gpx problem
* the ariege json is the same problem as in #1040

**Comment by zarov** [comment]
Lines are now visibles on the gpx path

**Comment by gchoqueux** [comment]
There's `non-coherent result 0 vs 1.` error  in example `globe_vector.html`.

**Comment by zarov** [comment]
So the last problem has been solved ! It  was due to the fact that `FileSource#urlFromExtent` was returning the same string without some unicity. This is now fixed with the generation of a random string.

**Comment by gchoqueux** [comment]
there is always the problem with `non-coherent result`.

**Comment by zarov** [comment]
The caching in `DataSourceProvider` has been reworked, and it fixes all the problem we got in a very simpler way.

File: src/Provider/DataSourceProvider.js
Comment (zarov) [inline src/Provider/DataSourceProvider.js:59]: We currently have some problems where the `parsedFile.extent` is in `4978`: this makes no sense (in itowns at least) to have a geocentric extent. This problem is not fixed in this PR, but an issue will be opened about it.

**Comment by gchoqueux** [comment]
there's a problem with immersive example...

**Comment by zarov** [comment]
Fixed in 5d38ecc but not very satisfied with it... WDYT ?

File: src/Provider/DataSourceProvider.js
Comment (gchoqueux) [reply src/Provider/DataSourceProvider.js:59]: could you open it?

**Review by gchoqueux — APPROVED** [review]
LGTM, Thanks for your effort.

**Comment by gchoqueux** [comment]
Rebase and edit commit 5d38ecc message before merge the PR.

File: src/Provider/DataSourceProvider.js
Comment (zarov) [reply src/Provider/DataSourceProvider.js:59]: Done in #1063

**Comment by zarov** [comment]
Thanks !

**Pull request merged**
